### ISSUE TITLE
Wrapping eval_env with HERGoalEnvWrapper when evaluating training using HER

### ISSUE THREAD

**Issue by PierreExeter** [hill-a/stable-baselines#893]
Hello,

I would like to discuss submitting a PR in relation to [this issue](https://github.com/araffin/rl-baselines-zoo/issues/77) from Stable Baselines Zoo. 

The [EvalCallback](https://github.com/hill-a/stable-baselines/blob/0c956eb9b10c9bd927331da62d26b85c8b5f0662/stable_baselines/common/callbacks.py#L222) does not wrap the evaluation environment with `HERGoalEnvWrapper` but rather in `DummyVecEnv`. This creates some inconsistency in the evaluation, see the issue mentioned above.

I propose to add this code [here](https://github.com/hill-a/stable-baselines/blob/0c956eb9b10c9bd927331da62d26b85c8b5f0662/stable_baselines/common/callbacks.py#L257)
```python
if not isinstance(eval_env, VecEnv) and not isinstance(eval_env, stable_baselines.her.HERGoalEnvWrapper):
    eval_env = DummyVecEnv([lambda: eval_env])
    assert eval_env.num_envs == 1, "You must pass only one environment for evaluation"
elif isinstance(eval_env, stable_baselines.her.HERGoalEnvWrapper):
    print("Not replacing HERGoalEnvWrapper env by a DummyVecEnv")
```

and this code [here](https://github.com/hill-a/stable-baselines/blob/0c956eb9b10c9bd927331da62d26b85c8b5f0662/stable_baselines/her/utils.py#L67) (as it is now necessary to flatten the observation list of list)
```python
if len(obs_dict[KEY_ORDER[0]].shape) == 2:
    # flatten 'observation', 'achieved_goal' and 'desired_goal'
    for key in KEY_ORDER:
        obs_dict[key] = obs_dict[key][0]
```

Please let me know what you think and whether I should submit this PR.

**Comment by araffin** [comment]
>I propose to add this code here

Why would we need to change the SB code, what you did in the zoo is sufficient, no?

**Comment by PierreExeter** [comment]
No I don't think it's sufficient. With my changes in the zoo, training with this 
```bash
python train.py --algo her --env parking-v0 -n 10000
```
still returns this warning
```bash
UserWarning: Training and eval env are not of the same type<stable_baselines.her.utils.HERGoalEnvWrapper object at 0x7fe4d129d668> != <stable_baselines.common.vec_env.dummy_vec_env.DummyVecEnv object at 0x7fe4c2e579e8>
  "{} != {}".format(self.training_env, self.eval_env))
```
The 1st change proposed gets rid of this warning by preventing the eval_env from being wrapped in a DummyVecEnv. I realized that the 2nd change is not necessary (it was throwing an error about observation dimensions but not anymore).

Anyway maybe these changes are not so important in the end. Also I know you're busy this week and I found this quick fix that works for me so we can leave this for later. Thanks.
